### DISCUSSION TITLE
Spreading Factor for RNode in Sideband?

### DISCUSSION THREAD

**Discussion by Starfall044** [markqvist/Reticulum#249] [Help & Questions]
I've got two LILYGO T-Beam 1.1-based RNodes talking to each other with Sideband, and it works great as with these settings: 

Freq: 867.2
BW: 62.5
TX: 14
Spreading Factor: 8
Coding Rate: 6

Announcing, sending messages, receiving messages, no problems. I can change the TX power around with no issues, but if I set SF to 9, for instance, my messages all fail. The weird thing is, I can still "announce" and see the other device pick up the new announce. But when I try to send a message, it tries for a while and then turns red. 

While I'm chasing down options, figured I'd throw it out to the community: Are there only certain combinations of variables that will work? Or does the T-Beam only support certain spreading factors? Why would this be happening?

Thanks in advance!

**Comment by markqvist** [comment]
That's odd. Definitely should work. The resulting bitrate at SF9 intsead of SF8 is 732 bits per second, which should be fine.

**Reply by pizdjuk** [reply]
same problem here.

**Comment by Ilomantis** [comment]
I have the same problem too, albeit it seems to start playing up once SF is >10. The activity graph on the rnode also appears to get stuck or freeze when sending.
I am using LilyGO LoRa32 2.1 boards.

**Reply by markqvist** [reply]
The activity graph will always stop when sending. It can't read the spectrum activity when the transceiver is sending ;)

**Comment by pizzadoglover911** [comment]
I am running into this problem at the moment, though I cannot tell if the tx pwr, spread, or coding rate is the source of the trouble. 

This showed itself on my Heltec boards while I was trying to connect through sideband. Bluetooth or USB, the app launch would not "latch" onto the RNode. Lowered all the settings closer to the stock config(from documentation? i forget) and sideband will communicate with the Rnode again. 

I will play around with the options to see what is causing it. Again this is on a Heltec board and two seperate models of phones.

EDIT: 
With the following: 915mhz, 125.0khz band, 8 spread, and 5 coding rate, I could only increase the TX power to 17dbm. The documentation for SX1276 indicates 20 dbm or 100mw. Maybe I am misunderstanding that limit, or is this an issue with sideband or my config of? 
At 17dmn and the above settings, Sideband interfaces with the RNodes via bluetooth or usb without issue.

**Reply by markqvist** [reply]
The RNode firmware supports a max TX power of 17dBM for the SX1276.

The documentation for the boards themselves is a bit misleading. The SX1276 chips can **only** do 20dBM output for *very* short bursts, without causing damage to the chip, so it is intentionally disabled by the firmware.

Maybe I should make this more clear somewhere.

**Reply by pizzadoglover911** [reply]
Ah I see. It would have helped if I read past the "key features" section(the first page) of the data sheet! Semtech lists +20dBm transmit power at only 1% duty cycle. 17dBm for constant duty. Nice!

> PA_HP (High Power), connected to the PA_BOOST pin, covers all frequency bands the chip addresses. It permits continuous operation at up to +17 dBm and duly cycled operation at up to +20dBm. For full details of operation at +20dBm please consult seclion 3.4.3

Bringing this back to the original topic, I increased spread and coding rate to 12 and 8 respectivley. I could view an annouce through another RNode, but messages would fail. This is at 17dBm. At spreadingfactor=10 and codingrate=7, messages were slower(as expected) but did transmit successfully. 

e: wanted to link this page as it was hard to come by for me and relevant https://unsigned.io/understanding-lora-parameters/

**Reply by markqvist** [reply]
>  I increased spread and coding rate to 12 and 8 respectivley. I could view an annouce through another RNode, but messages would fail.

Which is because you are now running at an effective bitrate of 183 bits per second, way lower than the absolute minimum supported by Reticulum, which is currently 475 bits per second ;)

To reach even lower bitrates, some adaptive timing techniques need to be added to Reticulum.

I have a handy calculator for these parameters available here, by the way: https://unsigned.io/articles/2022_05_05_understanding-lora-parameters.html

**Reply by pizzadoglover911** [reply]
> I have a handy calculator for these parameters available here, by the way: https://unsigned.io/articles/2022_05_05_understanding-lora-parameters.html

On the topic of your calculator, I am confused by how the coding rate is referenced. In .reticulum/.config, coding rate is noted as 5-8. On your calculator, it is expressed at 4:5, 4:6, 4:7, and 4:8. What is the significance of the 4?

Going to try more range testing today, using your calculator to achieve the minimum bitrate you described. 

EDIT: Answered my own question, quote from [thethingsnetwork](https://www.thethingsnetwork.org/docs/lorawan/fec-and-code-rate/):

> For example, if the code rate is 5/7, for every 5 bits of useful information, the coder generates a total of 7 bits of data, of which 2 bits are redundant.

**Comment by acehoss** [comment]
@markqvist Not to pile on, but this seems like it could be related to issues I mentioned while testing the Buffer example. When I was running 125/10/7 (0.7 Kbps), I noticed significant packet loss, up to 25%. It was good for testing the retry mechanism of Channel, but the RNodes were inches away from each other and should not have had any problem communicating. I tried power set to 17 and power set to 7, it acted the same either way. 

The receiving RNode would indicate DCD while the other was transmitting, but then no packet would be received.

Also at 62.5/10/6 to hit that 410 bps rate, I saw nearly 100% packet loss. The receiving RNode would indicate DCD but did not appear to receive a packet. I know I didn't have the tuning set to be able to run links at 410 bps, but almost none of the path requests or announces were received so I never really got to the link establishment phase of the connection at 410 bps.

**Reply by markqvist** [reply]
Thanks for adding to this. I have been experimenting a bit with it too, and have been seeing similar behaviour. This might actually be something in the firmware going sideways at those very low bitrates. I'll have to look into it more closely to determine the real cause.

If anyone discovers anything else in regards to this, please just post it here :)

**Reply by cobraPA** [reply]
I've seen notes that Lora devices at too close a proximity may interact badly...  Seeing acehoss say the devices are inches apart makes me wonder if this is a factor.  If you move them further away and/or reduce power (a lot) and the packet loss disappears, I would think you have proven the case.  My limited understanding is this is related to the simple antenna input stages / noise and/or filtering in these low cost Lora devices.  (Re-reading, I note you did lower power, so maybe the minimum power is still too high at that close proximity.)

**Comment by pizzadoglover911** [comment]
915mhz, 250khz band, 17dbm power, 12 SF, 6 coding rate. This setup wont communicate within arms reach of the two transceivers. The both Rnodes will display transmission activity, but the message will fail. The calculator shows 488bps at these settings. Too low still. 

Moving to a SF of 11 (rest the same as above) resulted in 895kbps and an effective connection. Back to the range testing.

**Reply by markqvist** [reply]
It *should* work down to around 475 bits per second, but "officially" the limit is 500 bps, so those settings *are* kinda pushing it ;)

Hopefully soon, there will be a more dynamic timing algorithm, that will allow even lower bitrates.

**Reply by pizdjuk** [reply]
I try the same setting (with SF=11). On Node A run  rncp -n -r. On node B rnpath <address from rncp>
and I get  Received invalid announce for <address from rncp>. 

And there are many other messages with the same schema: received invalid announce for <xxxx>

**Reply by LaneaLucy** [reply]
> Hopefully soon, there will be a more dynamic timing algorithm, that will allow even lower bitrates.

Do you have a timeframe for this feature?

**Comment by LaneaLucy** [comment]
Ok, something is broken: with SF 12 CR 5 BW 125 I get announcements through, but no messages. From what I read, this is expected because not enough bps. BUT when I set SF 11, CR 5 BW 125, I don't get any messages + no more announcements too... ???????
It should be the other way around...
SF 11 should theoretically have enough bps to work, but nothing works. SF 12 shouldn't have enough bps to work, but it works half...

Can someone tell me, what's wrong? I only change the SF, nothing else. So why is it the opposite of what it should be?

**Reply by pizdjuk** [reply]
I tried SF=9 and it's very unreliable too. SF=10 does not work.
Try BW=250 or BW=500 to see what happens. With wider bandwidth it could became more stable

**Comment by SebastianObi** [comment]
I have also tried with several SF.
For me it works up to SF9.
SF 10 etc. also only announcements come through.
Probably the data is somehow corrupted.
The send and receive leds show activity, but reticlum receives no data.

I used the TTGO T3_V1.6.1 with 868MHz

**Reply by pizdjuk** [reply]
> I have also tried with several SF. For me it works up to SF9. SF 10 etc. also only announcements come through. Probably the data is somehow corrupted. The send and receive leds show activity, but reticlum receives no data.
> 
> I used the TTGO T3_V1.6.1 with 868MHz

Try to send messages in two directions with SF=9. In one direction it worked ok, but in other  very bad or didnt work at all.

**Comment by faragher** [comment]
So, since this seems to happen a bit, I did some tests. Based on the Announce and Echo examples, I threw together some test programs. I can post them if that'd be helpful, but I think they need more information, such as RSSI values, before it's useful. I just wanted to share my initial results. All tests at 915 MHz, 1 dBm, and a range of around ten feet. Some of these are directly noted in the thread as known good or known bad.

```
BW    SF    CR      Baud     Notes
125   11      5      537     NO-GO, Error 1 
125   10      5      977     GO - 0% packet loss
125    9      5     1760     GO - 0% packet loss
125   10      7      698     NO-GO, Error 2
125   12      6      ---     Out of Spec - Error 3
125   11      6      ---     NO-GO, Error 1
250   12      6      488     NO-GO, Error 2, Long delay and 50-75% packet loss on path requests
250   11      6      895     GO - 10% packet loss
62.5   8      6     1300     NO-GO, Error 4
62.5   9      6      732     GO 0% packet loss
62.5  10      5      488     NO-GO Error 1, 10-25% Path request loss
62.5   7      6     2280     GO 0% packet loss, Note 1
```




**Error 1:** Sender requests path, Receiver acknowledges request, but sender never receives the path. Test failed.

Note: Receiver system is part of MLAT ADS-B system and is Zulu/GMT, Sender is at local time. Outputs are representative and may not be the same packets.

Typical output, Receiver:
```
[2023-06-06 23:35:59] [Debug] Answering path request for <3644e18e80b0ab3422d2f3eb15393664> on RNodeInterface[RNode LoRa Interface], destination is local to this system
[2023-06-06 23:36:00] [Debug] Path request for <3644e18e80b0ab3422d2f3eb15393664> on RNodeInterface[RNode LoRa Interface]
[2023-06-06 23:36:00] [Debug] Answering path request for <3644e18e80b0ab3422d2f3eb15393664> on RNodeInterface[RNode LoRa Interface], destination is local to this system
[2023-06-06 23:36:01] [Debug] Path request for <3644e18e80b0ab3422d2f3eb15393664> on RNodeInterface[RNode LoRa Interface]
```

Typical output, Sender:
```
[2023-06-06 17:35:56] [Notice] Destination is not yet known. Requesting path...
[2023-06-06 17:35:57] [Notice] Destination is not yet known. Requesting path...
[2023-06-06 17:35:58] [Notice] Destination is not yet known. Requesting path...
[2023-06-06 17:35:59] [Notice] Destination is not yet known. Requesting path...
```

**Error 2:** Presents as Error 1, but a several second delay prior to first response to path request.

**Error 3:** Baud rate out of specification, expected failure. No response to path request.

**Error 4:** Baud rate in specification, unexpected failure. No response to path request.

**Note 1:** RNode jammed open at start of trial. Meter on physical unit fully filled, device not responding on port. After restarting script, node behaved properly. Logs scrolled offscreen from previous test (BW 62.5 SF 10 CR 5) and could not be recovered. Assuming improper shutdown from previous test.


**Intermission:** I'm going to post this for fear of technical problems, alter the timings and see if there's any difference in performance.

**Comment by faragher** [comment]
I keep transcribing results in these pages, and I keep losing them. I guess I'm the fool for not typing them up elsewhere first.

I increased the delay after each path request to 5 seconds. This improved 62.5/8/6, and 125/10-12/6&7 that were tested above. 11/6 had 20-30% losses, however. 125/11-12/6 were very slow, with 6 and 13 second cycles respectively. This implies the RNode buffers are quite full and active while the Python script is already terminated. Further testing will continue after this post.

Path Request: 51 bytes
Announce: 167 bytes
Test Packet: 35 bytes (2 header, 16 address, 1 context, 16 payload)

Given 244/448 baud:

A path request would take 1.67/0.83 seconds.

A complete test cycle (218 bytes) would take 7.4/3.8 seconds. These don't line up with the cycle times I was getting 1:1, but it does show the importance of Python side timing. I can only assume the 1 second path request delay failed because it got back a request with a different timestamp or something similar, as when N returned N+1 had already been sent.

To the original question, of these 62.5/10/5 is the only one I can't get to work with raw Reticulum, and I think it's probably workable with just a few timing tweaks, but I think that misses the more universal solution. The RNode interface knows its parameters and performance and computes its own bitrate. Timing of programs and timeouts using the RNode should be based on its actual settings, not a best-fit solution. By threading the interface routines, each specific interface could use their most appropriate timings without unduely affecting the main program. 

Of course, for all I know Sideband does this already, but it's good to know, from a Reticulum standpoint, and from a LilyGo RNode standpoint, these all appear to be functional, if treated correctly.

Although the speed is something to consider. A path request taking almost a second and a simple announce taking over two was more than I planned for. I need to seriously plan out bandwidth for projects going forward.

**Comment by markqvist** [comment]
Thanks for doing these tests @faragher, very helpful! It would be great to post the test programs. I could use them for debugging this as well. If you want to add RSSI and SNR metrics before sharing the scripts, that would be absolutely awesome.

Reticulum adds SNR and RSSI metrics to every packet instance, if the underlying interface supports it. That means, that if your packets are coming from an RNode interface, you can get the stats simply from `packet.rssi` and `packet.snr`. You can also get the last receive RSSI and SNR directly from the `RNodeInterface` instance by reading the `r_stat_rssi` and `r_stat_snr` properties.

**Reply by SebastianObi** [reply]
Interesting to get directly these parameters.
There is a small bug with a missing init of r_stat_snr.
I made a pull request.

**Reply by SebastianObi** [reply]
I misunderstood the function of these values. I thought they could be queried by the interface at any time. But that's not the case. After the data has been passed to the transport, they are set back to "none".
So it is better to get the rssi values from the packet.
```
def processIncoming(self, data):
        self.rxb += len(data)
        self.owner.inbound(data, self)
        self.r_stat_rssi = None
        self.r_stat_snr = None
```

BTW, the rssi and snr are only on packets. Would there be a possibility to query these values for a received anounces? The announce callback does not pass the packet. Maybe this is some usefull feature request?

**Reply by markqvist** [reply]
Ah, you're right, I forgot about that. How to access it globally will likely need thinking through a bit more...

**Reply by faragher** [reply]
You know how test scripts are. I'll polish things up and add some real guidance and post the scripts.

**Comment by LaneaLucy** [comment]
I now have new results. 
SF 12 CR 5 BW 125 only announcement work. 
SF 11 CR 5 BW 125 nothing works.
SF 10 CR 5 BW 125 works reliable only with nomadnet, sideband only some messages get received and transmitting only worked 1 time in 6h test

**Comment by markqvist** [comment]
Okay, I just ran quite a few tests here on various RNode hardware. All devices were using 2 dBm output power, antennas with 0 dBi gain, and were distanced about 1 meter from each other. Noise floor at operating frequency around -70 dBm.

I tested in three steps:

1. Whether announces work
2. Whether link establishment works
3. Whether LXMF communications work

So far I have **not** been able to reproduce any of the unexpected behaviour. The only caveat is, that on the narrower frequencies (usually below 41.7 KHz), a frequency deviation of more than about a KHz will usually introduce problems, and you will need to make sure, that devices are actually operating on the same frequency. If not, adjust their frequencies a bit, so they actually line up.

| BW | CR | SF | Bitrate | Announce | Link | LXMF Comms | Notes |
|-----|-----|-----|-----|-----|-----|-----|-----|
|7.8 KHz | 5 | 7 | 341 bps | ✔️ | ❌ | ❌| Under bitrate limit; not expected to work |
|10.4 KHz | 5 | 7 | 455 bps | ✔️ | ✔️ | ✔️| Under bitrate limit; not expected to work |
|15.6 KHz | 5 | 7 | 682 bps | ✔️ | ✔️ | ✔️| Expected to work |
|20.8 KHz | 5 | 8 | 520 bps | ✔️ | ✔️ | ✔️| Expected to work |
|20.8 KHz | 5 | 7 | 910 bps | ✔️ | ✔️ | ✔️| Expected to work |
| 31.25 KHz | 5 | 8 | 781 bps | ✔️ | ✔️ | ✔️| Expected to work |
| 31.25 KHz | 5 | 7 | 1.37 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 41.7 KHz | 5 | 9 | 586 bps | ✔️ | ✔️ | ✔️| Expected to work |
| 41.7 KHz | 5 | 8 | 1.04 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 41.7 KHz | 5 | 7 | 2.73 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 62.5 KHz | 5 | 9 | 897 bps | ✔️ | ✔️ | ✔️| Expected to work |
| 62.5 KHz | 5 | 8 | 1.56 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 62.5 KHz | 5 | 7 | 2.73 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
|125 KHz | 5 | 12 | 293 bps | ✔️ | ❌ | ❌ | Under bitrate limit; not expected to work|
|125 KHz | 5 | 11 | 537 bps | ✔️ | ✔️ | ✔️| Expected to work |
|125 KHz | 5 | 10 | 977 bps | ✔️ | ✔️ | ✔️| Expected to work |
|125 KHz | 5 | 9 | 1.76 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
|125 KHz | 5 | 8 | 3.13 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
|125 KHz | 5 | 7 | 5.47 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 250 KHz | 5 | 12 | 1.17 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 250 KHz | 5 | 11 | 2.15 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 250 KHz | 5 | 10 | 1.95 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 250 KHz | 5 | 9 | 3.52 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 250 KHz | 5 | 8 | 6.25 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 250 KHz | 5 | 7 | 10.94 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 500 KHz | 5 | 12 | 1.17 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 500 KHz | 5 | 11 | 2.15 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 500 KHz | 5 | 10 | 3.91 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 500 KHz | 5 | 9 | 7.03 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 500 KHz | 5 | 8 | 12.5 Kbps | ✔️ | ✔️ | ✔️| Expected to work |
| 500 KHz | 5 | 7 | 21.88 Kbps | ✔️ | ✔️ | ✔️| Expected to work |

It all checks out as it should, at least here in my end. Only surprise for me was, that 10.4 KHz, SF7, CR5 actually seemed to work, even though it shouldn't.

**Reply by pizdjuk** [reply]
Only to be sure: did you shut down the AutoInterface? As I did early tests I forget about it and got very  "good" results too xD

**Reply by faragher** [reply]
Judging by the failures, it would appear so, but it's a great thing to verify!

**Reply by markqvist** [reply]
Thanks for the reminder. But look, I wrote this thing. Of course I did.

**Comment by faragher** [comment]
This test script doesn't currently sweep values, but is designed to test a single interface setup.

https://github.com/faragher/RNode_Setting_Test

**Reply by markqvist** [reply]
That is great! Thanks so much for sharing this!
